File: packages/json/llm-json-stream-ts/packages/llm-json-stream/test/properties/string_property.test.ts
/**
 * Tests for string property parsing and streaming
 * Uses async iterators as the primary streaming interface.
 */

import { describe, expect, test } from "@jest/globals";
import { JsonStreamParser } from "../../src/classes/json_stream_parser.js";
import { streamTextInChunks } from "../../src/utilities/stream_text_in_chunks.js";

describe("String Property Tests", () => {
    test("simple string value", async () => {
        const json = '{"name":"Alice"}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 3,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const nameStream = parser.getStringProperty("name");

        // Accumulate stream chunks using async iterator
        const chunks: string[] = [];
        for await (const chunk of nameStream) {
            chunks.push(chunk);
        }

        const finalValue = await nameStream.promise;

        // Verify accumulated chunks form the complete value
        expect(chunks.join("")).toBe("Alice");
        expect(finalValue).toBe("Alice");
    });

    test("string with escape sequences - newline", async () => {
        const json = '{"text":"Hello\\nWorld"}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 4,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const textStream = parser.getStringProperty("text");

        const chunks: string[] = [];
        for await (const chunk of textStream) {
            chunks.push(chunk);
        }

        const finalValue = await textStream.promise;

        expect(chunks.join("")).toBe("Hello\nWorld");
        expect(finalValue).toBe("Hello\nWorld");
    });

    test("string with escape sequences - tab", async () => {
        const json = '{"text":"Hello\\tWorld"}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 3,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const textStream = parser.getStringProperty("text");

        const finalValue = await textStream.promise;
        expect(finalValue).toBe("Hello\tWorld");
    });

    test("string with escape sequences - quotes", async () => {
        const json = '{"text":"She said \\"Hello\\""}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 5,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const textStream = parser.getStringProperty("text");

        const finalValue = await textStream.promise;
        expect(finalValue).toBe('She said "Hello"');
    });

    test("string with escape sequences - backslash", async () => {
        const json = '{"path":"C:\\\\Users\\\\Name"}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 6,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const pathStream = parser.getStringProperty("path");

        const finalValue = await pathStream.promise;
        expect(finalValue).toBe("C:\\Users\\Name");
    });

    test("empty string", async () => {
        const json = '{"value":""}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 4,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const valueStream = parser.getStringProperty("value");

        const finalValue = await valueStream.promise;
        expect(finalValue).toBe("");
    });

    test("string with unicode characters", async () => {
        const json = '{"emoji":"Hello 👋 World 🌍"}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 8,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const emojiStream = parser.getStringProperty("emoji");

        const finalValue = await emojiStream.promise;
        expect(finalValue).toBe("Hello 👋 World 🌍");
    });

    test("long string chunked across multiple stream chunks", async () => {
        const longText =
            "This is a very long string that will definitely be split across multiple chunks when streamed";
        const json = `{"description":"${longText}"}`;
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 10,
            interval: 5,
        });

        const parser = new JsonStreamParser(stream);
        const descStream = parser.getStringProperty("description");

        const chunks: string[] = [];
        for await (const chunk of descStream) {
            chunks.push(chunk);
        }

        const finalValue = await descStream.promise;

        // Should have received multiple chunks
        expect(chunks.length).toBeGreaterThan(1);
        expect(chunks.join("")).toBe(longText);
        expect(finalValue).toBe(longText);
    });

    test("string property streaming - multiple chunks", async () => {
        const json = '{"title":"My Great Blog Post"}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 5,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const titleStream = parser.getStringProperty("title");

        const chunks: string[] = [];
        for await (const chunk of titleStream) {
            chunks.push(chunk);
        }

        const finalValue = await titleStream.promise;

        // The string should be emitted in chunks as it's parsed
        expect(chunks.length).toBeGreaterThan(0);
        expect(chunks.join("")).toBe("My Great Blog Post");
        expect(finalValue).toBe("My Great Blog Post");
    });

    test("nested string property", async () => {
        const json = '{"user":{"name":"Bob"}}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 6,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const nameStream = parser.getStringProperty("user.name");

        const finalValue = await nameStream.promise;
        expect(finalValue).toBe("Bob");
    });

    test("async iterator pattern with for await...of", async () => {
        const json = '{"message":"Hello World from LLM"}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 5,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);
        const messageStream = parser.getStringProperty("message");

        // Accumulate chunks using modern async iterator
        const chunks: string[] = [];
        for await (const chunk of messageStream) {
            chunks.push(chunk);
        }

        // Verify accumulated chunks form the complete value
        const result = chunks.join("");
        expect(result).toBe("Hello World from LLM");
        expect(chunks.length).toBeGreaterThan(1); // Ensure we got multiple chunks
    });

    test("async iterator with non-streaming types yields stringified value once", async () => {
        const json = '{"count":42,"active":true,"value":null}';
        const stream = streamTextInChunks({
            text: json,
            chunkSize: 5,
            interval: 10,
        });

        const parser = new JsonStreamParser(stream);

        // Number type - emits the value once
        const countStream = parser.getNumberProperty("count");
        const countValues: number[] = [];
        for await (const value of countStream) {
            countValues.push(value);
        }
        expect(countValues).toEqual([42]);
        expect(await countStream.promise).toBe(42);

        // Boolean type - emits the value once
        const activeStream = parser.getBooleanProperty("active");
        const activeValues: boolean[] = [];
        for await (const value of activeStream) {
            activeValues.push(value);
        }
        expect(activeValues).toEqual([true]);
        expect(await activeStream.promise).toBe(true);

        // Null type - emits the value once
        const valueStream = parser.getNullProperty("value");
        const nullValues: null[] = [];
        for await (const value of valueStream) {
            nullValues.push(value);
        }
        expect(nullValues).toEqual([null]);
        expect(await valueStream.promise).toBeNull();
    });
});

File: packages/json/llm_json_stream_dart/test/properties/string_property_test.dart
import 'package:llm_json_stream/llm_json_stream.dart';
import 'package:test/test.dart';

import 'list_property_test.dart';

/// Enable verbose logging to debug test execution
const bool verbose = false;

// TODO: HANDLE ESCAPE SEQUENCES IN STRINGS MORE CLEARLY
// I am unsure if the test expectations should have the actual escape characters interpreted
// (e.g., \n as newline) or kept as literal sequences (e.g., \n as backslash + n).
// Otherwise, the tests are passing but may not reflect intended behavior.

void main() {
  group('String Property Tests', () {
    test('simple string value', () async {
      if (verbose) print('\n[TEST] Simple string value');

      final json = '{"name":"Alice"}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final nameStream = parser.getStringProperty("name");

      // Accumulate stream chunks
      final chunks = <String>[];
      nameStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      // Wait for completion
      final finalValue = await nameStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      // Verify accumulated chunks form the complete value
      expect(chunks.join(''), equals('Alice'));

      // Verify future resolves to complete value
      expect(finalValue, equals('Alice'));
    });

    test('string with escape sequences - newline', () async {
      if (verbose) print('\n[TEST] String with newline escape');

      // JSON with \n escape sequence should produce an actual newline character
      final json = '{"text":"Hello\\nWorld"}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final textStream = parser.getStringProperty("text");

      final chunks = <String>[];
      textStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await textStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('Hello\nWorld'));
      expect(finalValue, equals('Hello\nWorld'));
    });

    test('string with escape sequences - tab', () async {
      if (verbose) print('\n[TEST] String with tab escape');

      // JSON with \t escape sequence should produce an actual tab character
      final json = '{"text":"Hello\\tWorld"}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final textStream = parser.getStringProperty("text");

      final chunks = <String>[];
      textStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await textStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('Hello\tWorld'));
      expect(finalValue, equals('Hello\tWorld'));
    });

    test('string with escaped quotes inside', () async {
      if (verbose) print('\n[TEST] String with escaped quotes');

      // JSON with \" escape sequences should produce actual quote characters
      final json = '{"quote":"She said \\"Hi\\""}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 5,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final quoteStream = parser.getStringProperty("quote");

      final chunks = <String>[];
      quoteStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await quoteStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('She said "Hi"'));
      expect(finalValue, equals('She said "Hi"'));
    });

    test('string with backslashes', () async {
      if (verbose) print('\n[TEST] String with backslashes');

      final json = '{"path":"C:\\\\Users\\\\file"}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final pathStream = parser.getStringProperty("path");

      final chunks = <String>[];
      pathStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await pathStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('C:\\Users\\file'));
      expect(finalValue, equals('C:\\Users\\file'));
    });

    test('empty string', () async {
      if (verbose) print('\n[TEST] Empty string');

      final json = '{"empty":""}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final emptyStream = parser.getStringProperty("empty");

      final chunks = <String>[];
      emptyStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await emptyStream.future.withTestTimeout();
      if (verbose) print('[FINAL] "$finalValue"');

      expect(chunks.join(''), equals(''));
      expect(finalValue, equals(''));
    });

    test('string with Unicode/emoji', () async {
      if (verbose) print('\n[TEST] String with Unicode');

      final json = '{"emoji":"👋🌍"}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 3,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final emojiStream = parser.getStringProperty("emoji");

      final chunks = <String>[];
      emojiStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await emojiStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('👋🌍'));
      expect(finalValue, equals('👋🌍'));
    });

    test('nested string access with dot notation', () async {
      if (verbose) print('\n[TEST] Nested string access');

      final json = '{"user":{"name":"Bob"}}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 4,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final nameStream = parser.getStringProperty("user.name");

      final chunks = <String>[];
      nameStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await nameStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('Bob'));
      expect(finalValue, equals('Bob'));
    });

    test('deeply nested string access', () async {
      if (verbose) print('\n[TEST] Deeply nested string access');

      final json = '{"auth":{"user":{"profile":{"name":"Charlie"}}}}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 5,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final nameStream = parser.getStringProperty("auth.user.profile.name");

      final chunks = <String>[];
      nameStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await nameStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('Charlie'));
      expect(finalValue, equals('Charlie'));
    });

    test('string with multiple escape sequences', () async {
      if (verbose) print('\n[TEST] String with multiple escapes');

      final json = '{"msg":"Line1\\nLine2\\tTabbed\\r\\nLine3"}';
      if (verbose) print('[JSON] $json');

      final stream = streamTextInChunks(
        text: json,
        chunkSize: 6,
        interval: Duration(milliseconds: 10),
      );
      final parser = JsonStreamParser(stream);

      final msgStream = parser.getStringProperty("msg");

      final chunks = <String>[];
      msgStream.stream.listen((chunk) {
        if (verbose) print('[CHUNK] "$chunk"');
        chunks.add(chunk);
      });

      final finalValue = await msgStream.future.withTestTimeout();
      if (verbose) print('[FINAL] $finalValue');

      expect(chunks.join(''), equals('Line1\nLine2\tTabbed\r\nLine3'));
      expect(finalValue, equals('Line1\nLine2\tTabbed\r\nLine3'));
    });
  });
}


